### PR TITLE
Allow to renew tokens

### PR DESCRIPTION
**Pull request by Quentin01**

### CODE REVIEW

File: lib/errors/cancel.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+'use strict';
+
+var util = require('util');
+
+var CancelError = function() {
+  this.name = 'CancelError';
+  this.message = "User have canceled the authorization";
```
Comment (Neamar) [inline lib/errors/cancel.js]: User cancelled the authorization

File: lib/errors/token.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
+'use strict';
+
+var util = require('util');
+
+var TokenError = function() {
+  this.name = 'TokenError';
+  this.message = "Token need to be refresh";
```
Comment (Neamar) [inline lib/errors/token.js]: Token is not valid anymore

File: lib/handlers/init/callback.js [hunk 3]
```diff
@@ -60,15 +60,21 @@ module.exports.getGenerator = function getGenerator(retrieveTokens, config) {
...
       },
-      function createNewToken(accountName, providerData, tempToken, anyfetchAccessToken, cb) {
+      function createToken(accountName, providerData, tempToken, anyfetchAccessToken, cb) {
         // Create new token
         var token = new Token({
           anyfetchToken: anyfetchAccessToken,
           data: providerData,
-          accountName: accountName
+          accountName: accountName,
+          needRefresh: false
```
Comment (Neamar) [inline lib/handlers/init/callback.js]: requireRefresh

File: lib/handlers/init/callback.js [hunk 4]
```diff
@@ -60,15 +60,21 @@ module.exports.getGenerator = function getGenerator(retrieveTokens, config) {
...
         });
 
-        token.save(rarity.carryAndSlice([anyfetchAccessToken, tempToken], 3, cb));
+        var upsertData = token.toObject();
+
+        delete upsertData.lastUpdate;
+        delete upsertData._id;
+
+        Token.update({anyfetchToken: tempToken.anyfetchToken || anyfetchAccessToken}, upsertData, {upsert: true}, rarity.carryAndSlice([anyfetchAccessToken, tempToken], 3, cb));
```
Comment (Neamar) [inline lib/handlers/init/callback.js]: // Update previous token, or insert new one

Comment (Neamar) [inline lib/handlers/init/callback.js]: // For cases where token is invalid

File: lib/models/temp-token.js [hunk 5]
```diff
@@ -10,6 +10,9 @@ var TempTokenSchema = new mongoose.Schema({
   // Authorization code to communicate with AnyFetch
   anyfetchCode: '',
 
+  // If it's a new token or we renew another token
```
Comment (Neamar) [inline lib/models/temp-token.js]: If it's a new token or another token renewal

File: lib/models/token.js [hunk 6]
```diff
@@ -23,7 +23,10 @@ var TokenSchema = new mongoose.Schema({
   isUpdating: {type: Boolean, default: false},
 
   // Date we started last update
-  lastUpdate: {type: Date, default: null}
+  lastUpdate: {type: Date, default: null},
+
+  // Need refresh
```
Comment (Neamar) [inline lib/models/token.js]: // Is this token is stale and requiring a refresh?

File: lib/models/token.js [hunk 7]
```diff
@@ -23,7 +23,10 @@ var TokenSchema = new mongoose.Schema({
   isUpdating: {type: Boolean, default: false},
 
   // Date we started last update
-  lastUpdate: {type: Date, default: null}
+  lastUpdate: {type: Date, default: null},
+
+  // Is this token is stale and requiring a refresh?
```
Comment (Neamar) [inline lib/models/token.js]: Is this token ~~is~~ stale

**Comment by Neamar** [comment]
Let's add a test for this?

**Comment by Quentin01** [comment]
I can test some things, but can't test oauth with renewal.

File: test/handlers/init/callback.js [hunk 8]
```diff
@@ -14,13 +14,14 @@ var updateFile = helpers.updateFile;
...
   before(function createMockServer(done) {
-    var server = AnyFetch.createMockServer();
+    mockServer = AnyFetch.createMockServer();
 
     var port = 1337;
     var apiUrl = 'http://localhost:' + port;
 
-    server.listen(port, function() {
+    mockServer.listen(port, function() {
       console.log('AnyFetch mock server running on ' + apiUrl);
```
Comment (Neamar) [inline test/handlers/init/callback.js]: do we need the log?

File: test/handlers/update.js [hunk 9]
```diff
@@ -33,6 +35,26 @@ describe("POST /update endpoint", function() {
...
 
+  var mockServer;
+  before(function(done) {
+    mockServer = AnyFetch.createMockServer();
+
+    var port = 1337;
+    var apiUrl = 'http://localhost:' + port;
+
+    mockServer.listen(port, function() {
+      console.log('AnyFetch mock server running on ' + apiUrl);
```
Comment (Neamar) [inline test/handlers/update.js]: do we need the log?

File: test/handlers/update.js [hunk 10]
```diff
@@ -274,4 +326,46 @@ describe("POST /update endpoint", function() {
...
+          cb(err);
+        });
+
+        updateServer(server, null, function(err) {
+          if(err) {
+            return cb(err);
+          }
+
+          server.usersQueue.once('empty', function() {
+            console.log("EMPTY");
```
Comment (Neamar) [inline test/handlers/update.js]: do we need the log?

**Pull request merged**